## Problem #67: Maximum Path Sum II
[Link to Problem](https://projecteuler.net/problem=67)

### Problem Description

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

$$

\bold{3} \\[5pt]
\bold{7} \quad 4 \\[5pt]
2 \quad \bold{4} \quad 6 \\[5pt]
8 \quad 5 \quad \bold{9} \quad 3

$$

That is, $3 + 7 + 4 + 9 = 23$.

Find the maximum total from top to bottom in [67_triangle.txt](67_triangle.txt), a 15K text file containing a triangle with one-hundred rows.

**NOTE:** This is a much more difficult version of [Problem 18](../001%20-%20050/018.py). It is not possible to try every route to solve this problem, as there are $2^{99}$ altogether! If you could check one trillion ($10^{12}$) routes every second it would take over twenty billion years to check them all. There is an efficient algorithm to solve it. ;o)

### Approach

Basic **dynamic programming**.

For row $i$ and column $j$, let $\text{dp}[i][j]$ = the maximum sum we can achieve if we stop at cell $(i, j)$.

Recurence: $dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - 1])$

In [1]:
dp = [[]]

with open('67_triangle.txt', 'r') as f:
    dp = [[int(x) for x in line.split()] for line in f.read().split('\n')]

for i in range(1, len(dp)):
    for j in range(i + 1):
        prv = 0
        if j < i:
            prv = dp[i - 1][j]
        if j > 0:
            prv = max(prv, dp[i - 1][j - 1])
        dp[i][j] += prv

print(max(dp[-1]))

7273


###### Result: **7273** | Execution time: 0s

### Complexity analysis

$O(\text{Number of cells}) = O(N^2)$, where $N$ = size of the pyramid

##### Tags: #dynamic-programming